### DISCUSSION TITLE
10 luglio 2023 - esercizio 1

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/basi-di-dati-1#41] [Esami]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/40034429/538a1410-6469-436c-83c2-9b3eb78e07ae)

**Comment by Warcophyr** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/1ab9f0a6-a53d-4021-b0e9-fb5b0514bc44)

**Reply by FrascaHoryzont** [reply]
La prima è giusta ma la seconda penso sia meglio che fai la differenza alla fine, perché se prendi i prestiti che non sono stati fatti da clienti di Roma o Viterbo e ne ricavi il risultato, poi non sai se nei prestiti che hai tolto ci stanno libri degli stessi autori presi da clienti di Roma o Viterbo. Quindi il risultato secondo me è diverso.

**Reply by Warcophyr** [reply]
"poi non sai se nei prestiti che hai tolto ci stanno libri degli stessi autori presi da clienti di Roma o Viterbo" allora poiché nel punto 2a) chiede i libri mai presi da clienti da Roma o Viterbo come ho scritto tramite lo schema Prestito che ha attributo idCopia rimuovo i libri i quali sono stati presi da clienti di Roma o di Viterbo quindi si sei sicuro che li hai libri presi in prestito solo da clienti solo non di Roma o Viterbo (potrei non aver capito nel caso la tua domanda scusa in anticipo)

**Reply by FrascaHoryzont** [reply]
No scusa ho riletto adesso la consegna dell'esercizio. Nella mia domanda pensavo che l'esercizio chiedesse gli **autori** di libri mai presi in prestito da clienti di roma e viterbo. Ma siccome l'esercizio chiede i **libri** e non nello specifico gli autori allora l'esercizio è giusto come lo hai fatto. Scusa in anticipo.

**Comment by Frisbon** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/84731341/ef3e29d1-eba4-4d7d-b827-0d6febfaaa2b)

**Comment by CiottoloMaggico** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/79638284/8ca3874b-80f6-4b96-a470-db93de493dba)

**Comment by gioele-m-l** [comment]
> Ogni suggerimento o correzione di eventuali errori è ben accetto/a

1a)
1. Trovo i ***Prestiti*** di interesse, ovvero quelli con _DataP_ precedenti a Maggio 2023 e con _DataR_ equivalente a "00/00/0000":

$$PrestitiInteresse = \sigma_{DataP<01/05/2023 \land DataR=00/00/0000}(Prestito)$$

2. Il risultato è la proiezione sui dati dei ***Clienti***, l'attributo _Titolo_ di ***Libro*** e l'attributo _Id_ di ***Copia*** dei theta join tra le tabelle ***Cliente***, ***PrestitiInteresse***, ***Copia*** e ***Libro***:

$$Result = \pi_{Cliente.*, Libro.Titolo, Copia.Id}(Cliente \bowtie_{Cliente.Id=PrestitiInteresse.IdCliente} PrestitiInteresse \bowtie_{PrestitiInteresse.IdCopia=Copia.Id} Copia \bowtie_{Copia.IdLibro=Libro.Id} Libro)$$

1b)
1. Trovo i dati dei ***Clienti*** di "Roma" o "Viterbo" e ne proietto il _CF_ e l'_ID_:

$$ClientiRoV = \pi_{CF, ID}(\sigma_{Città='Roma' \lor Città='Viterbo'} (Cliente))$$

2. Cerco i ***Libri*** non validi, ovvero quei ***Libri*** per cui esiste una ***Copia*** coinvolta in almeno un ***Prestito*** richiesto dai ***Clienti*** in ***ClientiRoV***, successivamente proietto solo sui dati dei ***Libri***:

$$LibriInvalid = \pi_{Libro.*}(ClientiRoV \bowtie_{ClientiRoV.ID=Prestito.IdCliente} Prestito \bowtie_{Prestito.IdCopia=Copia.Id} Copia \bowtie_{Copia.IdLibro=Libro.Id} Libro)$$

3. Ora per ottenere i ***Libri*** che ci interessano basta effettuare la differenza tra la tabella ***Libro*** e la tabella ***LibriInvalid***:

$$LibriValid = Libri - LibriInvalid$$

4. Per finire effettuo una proiezione sugli attributi _Titolo_ e _Anno_ di ***Libro*** e tutti gli attributi di ***Autore*** dopo aver fatto il theta join tra le tabelle ***LibriValid*** e ***Autore***:

$$Result = \pi_{Libro.Titolo, Libro.Anno, Autore.*}(LibriValid \bowtie_{LibriValid.Autore=Autore.Id} Autore)$$
